src/internal/cache/redisclient.go

```diff
@@ -0,0 +1,145 @@
+package cache
+
+import (
+	"fmt"
+	"github.com/FZambia/sentinel"
+	"github.com/goharbor/harbor/src/common/utils/log"
+	"github.com/gomodule/redigo/redis"
+	"net/url"
+	"regexp"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+var knownPool sync.Map
+var m sync.Mutex
+var redisPathDBRegexp = regexp.MustCompile(`/(\d*)\z`)
+
+// RedisPoolParam ...
+type RedisPoolParam struct {
+	PoolMaxIdle     int
+	PoolMaxActive   int
+	PoolIdleTimeout time.Duration
+
+	DialConnectionTimeout time.Duration
+	DialReadTimeout       time.Duration
+	DialWriteTimeout      time.Duration
+}
+
+// GetRedisPool get a named redis pool
+// supported schema
+// redis://user:pass@redis_host:port/db
+// redis+sentinel://user:pass@redis_sentinel1:port1,redis_sentinel2:port2/monitor_name/db
+func GetRedisPool(name string, rawurl string, param *RedisPoolParam) (*redis.Pool, error) {
+	if p, ok := knownPool.Load(name); ok {
+		return p.(*redis.Pool), nil
+	}
+	m.Lock()
+	defer m.Unlock()
+	if p, ok := knownPool.Load(name); ok {
+		return p.(*redis.Pool), nil
+	}
+
+	u, err := url.Parse(rawurl)
+	if err != nil {
+		return nil, fmt.Errorf("bad redis url: %s", err)
+	}
+
+	if param == nil {
+		param = &RedisPoolParam{
+			PoolMaxIdle:           0,
+			PoolMaxActive:         1,
+			PoolIdleTimeout:       time.Minute,
+			DialConnectionTimeout: time.Second,
+			DialReadTimeout:       time.Second,
+			DialWriteTimeout:      time.Second,
+		}
+	}
+
+	fmt.Println("get redis pool:", rawurl)
+	if u.Scheme == "redis" || u.Scheme == "rediss" {
```

Comment:
Suggestion: you can define the logic for creating `redis` or `redis+sentinel` connection pools into different helper/utility functions for code well-organized purposes.